File: solutions/34. Find First and Last Position of Element in Sorted Array.java
//Time complexity: log(n)
// here we are first find the target element in the entire array then from there we are first moving in the left side for //finding the first occurance of the target, then we are moving towards right side for finding the last occurance of the //target.
//What to do while while moving towards left: again apply binary search and find the first occurance
//What to do while while moving towards right: again apply binary search and find the last occurance(this binary search //we need to start from first occurance of target till end of the array).
​
​
class Solution {
  public static int[] searchRange(int[] nums, int target) {
​
        int low=0;
        int high=nums.length-1;
        int[] ans=new int[2];
​
        if (nums.length == 0) {
            return new int[]{-1, -1};
        }else{
            int firstmin=firstmin(nums,low,high,target);
​
            if(firstmin==-1){
                return new int[]{-1,-1};
            }else{
                int lastmin=lastmin(nums,firstmin,high,target);
​
                return new int[]{firstmin,lastmin};
            }
        }
    }
​
    public static int firstmin(int[] arr, int low, int high, int target){
        while (low<=high){
            int mid=low+(high-low)/2;
​
            if(arr[mid] == target) {
                if (mid == 0) {
                    return mid;
                }else {
                    if (arr[mid-1]<target) {
                        return mid;
                    }else{
                        high=mid-1;
                    }
                }
            }else if (arr[mid]<target) {
                low=mid+1;
            }else {
                high=mid-1;
            }
        }
        return -1;
    }
​
    public static int lastmin(int[] arr, int low, int high, int target){
        while (low<=high){
            int mid=low+(high-low)/2;
​
            if (arr[mid] == target) {
                if (mid== arr.length-1 || (mid<arr.length-1 && arr[mid+1]>target)) {
                    return mid;
                }else{
                    low=mid+1;
                }
            }else if (arr[mid]<target) {
                low=mid+1;
            }else {
                high=mid-1;
            }
        }
        return -1;
    }
}

File: 34_FindFirstAndLastPositionOfElementInSortedArray.py
"""
Leetcode - https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/submissions/ (submitted)
TC - O(logN), SC- O(1)
Challenge - coming up with the logic in log(N)
Lecture - https://www.youtube.com/watch?v=LEq2VfGnNzI&ab_channel=%7BS30%7D

Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.
If target is not found in the array, return [-1, -1].
You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [5,7,7,8,8,10], target = 8
Output: [3,4]

Example 2:
Input: nums = [5,7,7,8,8,10], target = 6
Output: [-1,-1]

Example 3:
Input: nums = [], target = 0
Output: [-1,-1]

Constraints:
0 <= nums.length <= 10^5
-10^9 <= nums[i] <= 10^9
nums is a non-decreasing array.
-10^9 <= target <= 10^9
"""

'''
Idea - Two pass binary search. Find first element. From that index, do another binary search pass for last element.

Another approach will be a linear search in O(N) brute force
'''


class Solution:

    def searchRange(self, nums, target):
        firstIndex = self.binarySearch('first', nums, target, 0, len(nums) - 1)
        lastIndex = self.binarySearch('last', nums, target, firstIndex, len(nums) - 1)
        return [firstIndex, lastIndex]

    # TC - O(2*logN) = O(logN)
    def binarySearch(self, direction, arr, target, low, high):
        if low == -1:
            return -1
        resultIndex = -1
        while low <= high:
            mid = low + (high - low) // 2  # 2, 4
            if direction == 'first':
                if target <= arr[mid]:  # 8<=7, 8<=8
                    if arr[mid] == target:
                        resultIndex = mid  # 4
                    high = mid - 1  # 5
                else:
                    low = mid + 1  # 3
            else:
                if target >= arr[mid]:  # | 8 >= 8, 8 >= 8
                    if arr[mid] == target:
                        resultIndex = mid  # | 4
                    low = mid + 1  # | 4
                else:
                    high = mid - 1

        return resultIndex


nums = [5, 7, 7, 8, 8, 10]
target = 7
findRange = Solution().searchRange(nums, target)
print(findRange)
